### ISSUE TITLE
Global Per-Epoch Shuffling with TorchTrainer

### ISSUE THREAD

**Issue by MarylineChen** [ray-project/ray#47460]
### What happened + What you expected to happen

Hi! I'm using Ray with TorchTrainer and Ray Datasets to train a model. In my setup, I shuffle the dataset using `dataset.random_shuffle(seed=random.randint(1, 10000))` before passing it to TorchTrainer. However, I've noticed that the same seed is used at each epoch, resulting in the same shuffled order being applied across all epochs, despite my attempts to introduce randomness with different seeds.

Ideally, I would like the dataset to be reshuffled differently at the start of each epoch. One possible solution could involve using a mechanism such as `seed=(seed + trainer.current_epoch)` to generate a different shuffle for each epoch, but this doesn't seem possible with the current setup.

Is there currently a way to achieve global per-epoch shuffling with the existing Ray Dataset and TorchTrainer setup? If not, are there any plans to support this feature in the future? 

I would greatly appreciate any guidance or updates on this issue.

### Versions / Dependencies

```
ray==2.32.0
torch==2.3.1
pytorch-lightning==2.3.3 
```

### Reproduction script

```
from ray.data import Dataset, read_parquet
from ray.train.torch import TorchTrainer

def build_dataset(path, pipeline, seed) -> Dataset:
    dataset = read_parquet(path)
    dataset = dataset.map_batches(pipeline)
    dataset = dataset.random_shuffle(seed=seed)

train_dataset = build_dataset(train_path, pipeline, seed)
validation_dataset = build_dataset(valid_path, pipeline, seed)

trainer = TorchTrainer(
    train_loop_per_worker=train_func_per_worker,
    scaling_config=scaling_config,
    train_loop_config=train_config,
    run_config=run_config,
    datasets={
        "train": train_dataset,
        "validation": validation_dataset,
    },
)

result = trainer.fit()
```

### Issue Severity

Low: It annoys or frustrates me.

**Comment by haritsahm** [comment]
I've finally found another user experiencing the same issue. After extensive testing, I’ve confirmed that improper seeding has a significant negative impact on training stability. Even when following the documented seeding practices provided by Ray Train and Ray Data, I still encounter convergence issues, especially when using custom datasets that require custom samplers and batch samplers in a DDP (Distributed Data Parallel) setup. The model struggles to converge, even with extended training across more epochs.

One major challenge is implementing global per-epoch shuffling with Ray Data. To my knowledge, operations like `random_shuffle()` or `repartition()` require materializing the dataset in memory, which is not feasible for my use case due to the large dataset size.

Interestingly, when I switch back to using a standard PyTorch `Dataset` and `DataLoader`, the training stabilizes immediately. The model converges within the first few epochs, indicating that the issue is likely tied to how Ray handles data shuffling and seed reproducibility in distributed training scenarios.

**Comment by ywilke** [comment]
I’m running into the same issue, and it makes using Ray difficult for my project.

When training on a single node, I could simply call randomise_block_order() on my dataset at the start of each epoch to get a rough shuffle of the training data. However, since I’ve moved to multi-GPU training with TorchTrainer, I can no longer reshuffle the dataset between epochs.

The only available workaround seems to be using a local buffer shuffle, but in my experience that is too slow for practical use, which also aligns with what was reported in #42317.

I hope this limitation can be addressed in Ray, as proper per-epoch shuffling is crucial for many training workflows.

**Comment by arushi297** [comment]
I am facing the same issue. When I pass the same seed to the dataloader (local or global shuffle enabled), it generates the exact same data order for each epoch. I believe to handle it separately it would require me to recreate the dataloader everytime with different seed. Are there any solutions for this?
